android-json-form-wizard/src/main/java/com/vijay/jsonwizard/fragments/JsonWizardFormFragment.java

```diff
@@ -253,6 +263,121 @@ public void updateVisibilityOfNextAndSave(boolean next, boolean save) {
+                getJsonApi().setNextStep(next);
+                next();
+            }
+        }
+    }
+
+    /**
+     * Skips blank by relevance steps when previous is clicked on the json wizard forms.
+     */
+    public void skipStepOnPreviousPressed() {
+        if (skipBlankSteps()) {
+            JSONObject currentFormStep = getStep(getArguments().getString(JsonFormConstants.STEPNAME));
+            String next = currentFormStep.optString(JsonFormConstants.NEXT, "");
+            int currentFormStepNumber = getFormStepNumber(next);
+            for (int i = currentFormStepNumber; i >= 1; i--) {
+                JSONObject formStep = getJsonApi().getmJSONObject().optJSONObject(JsonFormConstants.STEP + i);
+                if (formStep != null) {
+                    checkIfStepIsBlank(formStep);
+                    if (shouldSkipStep()) {
+                        getFragmentManager().popBackStack();
+                    } else {
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * should not be used alone, use with {@link #nextStepHasNoSkipLogic()}
+     * Checks if a given step is blank due to relevance hiding all the widgets
+     *
+     * @param formStep {@link JSONObject}
+     */
+    private void checkIfStepIsBlank(JSONObject formStep) {
+        try {
+            if (formStep.has(JsonFormConstants.FIELDS)) {
+                JSONArray fields = formStep.getJSONArray(JsonFormConstants.FIELDS);
+                for (int i = 0; i < fields.length(); i++) {
+                    JSONObject field = fields.getJSONObject(i);
+                    if (field.has(JsonFormConstants.TYPE) && !JsonFormConstants.HIDDEN.equals(field.getString(JsonFormConstants.TYPE))) {
+                        boolean isVisible = field.optBoolean(JsonFormConstants.IS_VISIBLE, true);
+                        if (isVisible) {
+                            setShouldSkipStep(false);
+                            break;
+                        }
+                    }
+                }
+            }
+        } catch (JSONException e) {
+            Timber.e(e, "%s --> checkIfStepIsBlank", this.getClass().getCanonicalName());
+        }
+    }
+
+    /**
+     * Returns the current form step number when given than steps next step number.
+     * This number is used to figure out which steps to pop when previous is clicked.
+     *
+     * @param nextFormNumber {@link String}
+     * @return formNumber {@link Integer}
+     */
+    private int getFormStepNumber(String nextFormNumber) {
+        int formNumber = 0;
+        if (StringUtils.isNotBlank(nextFormNumber)) {
+            int currentFormNumber = Integer.parseInt(nextFormNumber.substring(4, 5)) - 1;
+            if (currentFormNumber > 0) {
+                formNumber = currentFormNumber;
+            } else if (currentFormNumber == 0) {
+                formNumber = 1;
+            }
+        }
+        return formNumber;
+    }
```

Comment:
Is this skip logic only relevant to wizard forms? Could it also be useful in other forms and in that case better placed in the `JsonFormFragment`?

Reply:
i mean skip logic is only relevant if there is more than one step .... Why would single step forms need a skip logic?

Reply:
`JsonFormFragment` is not solely for single step forms. 

`JsonFormFragment` is the parent to `JsonWizardFormFragment` so why not have the code there for apps that extend either fragment? 

We are currently using this functionality so this would be a breaking change and it's not clear that it's necessary to make this migration.

Reply:
I think that when two classes share a feature is when the super class should have some 'initializing' code with some customization on the child. In this case skipping step is only a feature for `JsonWizardFormFragment` hence i dont see the need for this. I would suggest you use `JsonWizardFormFragment` in your implementation since its a child class and will have all that `JsonFormFragment` has.

Reply:
I'm not sure I understand this. Why is skipping a step only unique to (or needed in) the `JsonWizardFormFragment` and not in the `JsonFormFragment`? Isn't this feature useful for all kinds of forms and not just wizard forms? 

It's not ideal to switch to `JsonWizardFormFragment` since we (and perhaps other implementations): 
1. Don't need the extra features it provides
1. Have custom a `JsonFormFragment` that has overridden behaviour from the parent class. 
1. Related to 2, have made some assumptions when working with `JsonFormFragment` which may be broken when migrating to `JsonWizardFormFragment`.

Reply:
**Question**.
How are you using the `skipping step feature` in a `JsonFormFragment` in your implementation? It cant be to skip step since you only have one step .... Am curious.

Reply:
I think there's some confusion around how `JsonFormFragment` works. `JsonFormFragment` does support multi-step forms. We have multiple multi-step forms in our implementation, [one of which](https://github.com/onaio/rdt-standard/blob/master/rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form-original.json) benefits from skipping blanks steps.

Reply:
To be clear you use `JsonFormFragment` in the above form??? if not could i see how you currently use `skipping step feature` in `JsonFormFragment` in your implementation.

Reply:
I use `JsonFormFragment` for all forms. https://github.com/onaio/rdt-standard/blob/a2cbe7073809d62cc978fef54d81030ae2333f8c/rdt-open-reader/src/covidIndonesia/assets/json.form/sample-collection-form-original.json#L5.  https://github.com/onaio/rdt-standard/blob/a2cbe7073809d62cc978fef54d81030ae2333f8c/rdt-open-reader/src/main/java/io/ona/rdt/fragment/RDTJsonFormFragment.java#L38.

Reply:
I think you should change that ... i.e you should use `JsonFormFragment` for single step forms and `JsonWizardFormFragment` for multi-step forms. 

By the way skipping step feature was in `JsonWizardFormFragment` before you added it to `JsonFormFragment`

Reply:
Is there a reason to use `JsonFormFragment` for single step forms only? I don't believe that's a requirement atm.

I moved it to `JsonFormFragment` because, again, it's useful for all types of forms, not just `JsonWizardFormFragment` forms. If another implementation wanted to extend `JsonFormFragment`, it could still benefit from the feature.

Reply:
Please also address why I should not be concerned about the three points I raised https://github.com/OpenSRP/opensrp-client-native-form/pull/407#discussion_r501510654.

Reply:
@githengi I'm not sure I follow, why can steps only be skipped on wizards?